railgun/internal/ctrlutils/utils.go

```diff
@@ -59,3 +60,24 @@ func IsAPIAvailable(mgr ctrl.Manager, obj client.Object) (bool, error) {
 
 	return true, nil
 }
+
+// HasAnnotation is a helper function to determine whether an object has a given annotation, and whether it's
+// to the value provided.
+func HasAnnotation(obj client.Object, key, val string) bool {
+	if v, ok := obj.GetAnnotations()[key]; ok {
+		if v == val {
+			return true
+		}
+	}
+	return false
```

Comment:
Branchless code is more readable and performant.

```suggestion
	v, ok := obj.GetAnnotations()[key]
	return ok && v == v.val
```

Reply:
So I got a little curious (just for the sake of curiosity) about whether or not the new solution was actually technically "more performant". I decided nerd out a little bit and take a couple minutes to break it down. I created the following program to represent the performance of the two different control flow mechanisms:

```golang
package main

var z bool
var x bool

func main() {}

func Batman() bool {
	if z {
		if x {
			return true
		}
	}
	return false
}

func Robin() bool {
	return z && x
}
```

And then I compiled it and determined the assembly code that would be produced by the Go compiler (`go1.16.3 linux/amd64`):

```shell
$ go build -buildmode=plugin
$ objdump -Mintel -d test.so > disassembled
```

This is the resulting (Intel syntax) Assembly code:

```asm
00000000000a3e80 <test.Batman>:
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e8d:	4c 8b 3d 94 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe94]        # 103d28 <test.x@@Base-0x305c6>
   a3e94:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e98:	75 07                	jne    a3ea1 <test.Batman+0x21>
   a3e9a:	c6 44 24 08 00       	mov    BYTE PTR [rsp+0x8],0x0
   a3ea0:	c3                   	ret
   a3ea1:	c6 44 24 08 01       	mov    BYTE PTR [rsp+0x8],0x1
   a3ea6:	c3                   	ret

00000000000a3ec0 <test.Robin>:
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3ecd:	4c 8b 3d 54 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe54]        # 103d28 <test.x@@Base-0x305c6>
   a3ed4:	41 0f b6 07          	movzx  eax,BYTE PTR [r15]
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
   a3edd:	31 c0                	xor    eax,eax
   a3ee0:	eb f6                	jmp    a3ed8 <test.Robin+0x18>
```

There are four cases to represent given the above assembly control flow:

1. z == false && x == true
2. z == false && x == false
3. z == true && x == false
4. z == true && x == true

The `Batman` function covers both `#1` and `#2` with the following code path running only 5 total instructions:

```asm
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e9a:	c6 44 24 08 00       	mov    BYTE PTR [rsp+0x8],0x0
   a3ea0:	c3                   	ret
```

Then `Batman` covers `#3` with:

```asm
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e8d:	4c 8b 3d 94 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe94]        # 103d28 <test.x@@Base-0x305c6>
   a3e94:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e98:	75 07                	jne    a3ea1 <test.Batman+0x21>
   a3e9a:	c6 44 24 08 00       	mov    BYTE PTR [rsp+0x8],0x0
   a3ea0:	c3                   	ret
```

And finally `#4` with:

```asm
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e8d:	4c 8b 3d 94 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe94]        # 103d28 <test.x@@Base-0x305c6>
   a3e94:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e98:	75 07                	jne    a3ea1 <test.Batman+0x21>
   a3ea1:	c6 44 24 08 01       	mov    BYTE PTR [rsp+0x8],0x1
   a3ea6:	c3                   	ret
```

Which is effectively the same as the solution for `#3`.

As for the `Robin` function when processing `#1` it carries out 7 total instructions:

```asm
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3ecd:	4c 8b 3d 54 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe54]        # 103d28 <test.x@@Base-0x305c6>
   a3ed4:	41 0f b6 07          	movzx  eax,BYTE PTR [r15]
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
```

The above implementation has to execute 4 `mov` instructions (including one zero-extending `movzx`) over the `Batman` function's solution which solves it with 2 `mov` instructions.


As for `#2` with `Robin`:

```asm
00000000000a3ec0 <test.Robin>:
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3edd:	31 c0                	xor    eax,eax
   a3ee0:	eb f6                	jmp    a3ed8 <test.Robin+0x18>
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
```

Similar to above 7 total instructions are needed (versus the 5 needed by the `Batman` function) including more writing to registers (using `xor`) to do the same job, and additional `jmp` instructions.

Problem `#3` is solved this way:

```asm
00000000000a3ec0 <test.Robin>:
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3ecd:	4c 8b 3d 54 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe54]        # 103d28 <test.x@@Base-0x305c6>
   a3ed4:	41 0f b6 07          	movzx  eax,BYTE PTR [r15]
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
```

And has to do an additional `movzx` over the `Batman` implementation, however it does reduce the number of comparison instructions by one. The solution for problem `#4` is effectively the same as `#3`, but again with the extra `movxz`.

Given all the above if we assume all input options are equal this means that the original solution requires two less instructions to solve the same problem and significantly less moving of bits between registers to do so as well. `Robin` technically has the advantage of using less memory, as it only requires 9 instructions to be loaded vs. 10. Overall this seems to indicate to me that **the original solution was ultimately "more performant"** from a runtime perspective _however there are more factors that could be taken into account_ such as seeing how branch prediction behavior would affect things, e.t.c.

I dug into this for fun (yes, I indeed find this kind of stuff fun :joy:) and because it doesn't cost much time to just compile, dump the binary and look at the resulting solutions in assembly, but I don't want to go further with this as I'm already swimming pretty deep in the waters of the pedantic ocean. It's pretty obvious when we're talking about a few bits in memory and a few extra instructions per call that we're going off the rails a bit as this is not 1981 and the systems which this software run have multiple millions and billions of processor cycles and memory availability that make this kind of thing beyond negligible, and I'm just fine with the second option, but I thought you might find it kinda interesting.

This is also just a shameless plug for my _secret hidden agenda_ that we *rewrite Kubernetes in assembly*!!! :scream: